### PR TITLE
fix(backup): remove empty string in secret backup

### PR DESCRIPTION
**Pull request by FrankYang0529**
**Problem:**
If there is an empty string in secret backup, we will get an invalid value error.

```
Invalid value: \"\": status.secretBackups.data.networkdata in body must be of type byte: \"\"
```

**Solution:**
Remove empty string in the secret backup.

**Related Issue:**
https://github.com/harvester/harvester/issues/1536

**Test plan:**

* Create a secret like the following:
```
apiVersion: v1
kind: Secret
metadata:
  name: secret-volume
type: Opaque
data:
  userdata: "I2Nsb3VkLWNvbmZpZwpwYXNzd29yZDogcGFzc3dvcmQKY2hwYXNzd2Q6CiAgZXhwaXJlOiBmYWxzZQpzc2hwd2F1dGg6IHRydWUKcGFja2FnZV91cGRhdGU6IHRydWUKcGFja2FnZXM6CiAgLSBxZW11LWd1ZXN0LWFnZW50CnJ1bmNtZDoKICAtIC0gc3lzdGVtY3RsCiAgICAtIGVuYWJsZQogICAgLSAnLS1ub3cnCiAgICAtIHFlbXUtZ2Euc2VydmljZQpzc2hfYXV0aG9yaXplZF9rZXlzOgogIC0gPi0KICAgIHNzaC1yc2EKICAgIEFBQUFCM056YUMxeWMyRUFBQUFEQVFBQkFBQUJnUUM5Nzc4OFNvb3BPamZ2YmZQbkF2MzVHcWJrVFhENFRoWWdVZFZhT2FBV2ErTzVoSFNnZ05RTDkvK2xCQ0RsWGI3SzEvSGlmOFJiUkdURnpxU3d6RDJWNVhpcWJXcFFEZlJsVFg4T0s4SjdqTWxUYU96TVc3SUFyMjJGSFRveVlHQTAxUGtWdXkxV2UvSzV6RmIrdWVsZEt2NEZSZ0NkcVlqaGdhaklVZEYxRWRlOTF6WEFtb3BuSExRenk3SmZIZi94MEJrVnp0S1N1Y2d0aW1PdUJzODNFR1dENnBGbnh2UmNQMXBITGl4TXI0d3FQU1lSUFVzeW5nL0tKejVUOFd2NGdTL0NtS09Kc3EydjNxNGpWVTREMWhuR3l5SXovS0JqejI3ckluRUlRbVpmMXpOTkdBemlHdWc0cXdwbS9INVMwbEtlQWRyTmlOWitwY1p2MGZORUVLa292TmkyWm9DRDJzeGJ0TTZOM3d6ODhVdkRpck95a2JpdmY2U0ZnWlVVQ29JS0xpMnh5Ym84bFBUUm05emx1QUZrVjJrQzgvUDl3bkR4SzY1U0hwL1R5bFYxeXlvakowY2o5QUpyTU93czQ0czlYTUJVWUpNSjAyMEZqbXB0R3NBT1pPT3k1Q0x4OXZlMk5iNnJGcktiOFE4NkJjdU12TUsrbHcvSnFPMD0KICAgIG5vYWguZ2lsZGVyc2xlZXZlQHN1c2UuY29tCg=="
  networkdata: ""
```

* Create a VM with the above secret:
```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  # ...
spec:
    # ...
    spec:
      # ...
      volumes:
      # ...
      - cloudInitNoCloud:
          networkDataSecretRef:
            name: secret-volume
          secretRef:
            name: secret-volume
        name: cloudinitdisk
```

* Setup NTP/S3 backup target
* Backup VM and check vmbackup can go into Ready state
* Try to restore vmbackup to new vm and check the vm can go into Running state.

### CODE REVIEW

File: pkg/controller/master/backup/backup.go [hunk 1]
```diff
@@ -180,31 +180,57 @@ func (h *Handler) getVolumeBackups(backup *harvesterv1.VirtualMachineBackup, vm
...
+			secretBackupMap[secretFullName] = harvesterv1.SecretBackup{
 				Name: secret.Name,
-				Data: secret.Data,
-			})
+				Data: data,
+			}
 		}
 	}
 
+	secretBackups := make([]harvesterv1.SecretBackup, 0, len(secretBackupMap))
```
Comment (w13915984028) [inline pkg/controller/master/backup/backup.go]: It looks the secretBackupMap is used to skip duplicated names, when yes,  we may consider to improve it:

`secretBackupMap := map[string]harvesterv1.SecretBackup{} `  -----> `map[string]int/bool`

in the loop,  
when volume.CloudInitNoCloud.UserDataSecretRef
  generate name, add name into map, add Data into  secretBackups 
 volume.CloudInitNoCloud.NetworkDataSecretRef
   generate name, check if in map: yes, continue;  no, add name into map, add Data into secretBackups 

finally, return `secretBackups `directly, no need to do following for loop  to generate it

Comment (FrankYang0529) [reply pkg/controller/master/backup/backup.go]: Good catch. Updated it. Thank you.

File: pkg/controller/master/backup/backup.go [hunk 2]
```diff
@@ -181,26 +181,56 @@ func (h *Handler) getVolumeBackups(backup *harvesterv1.VirtualMachineBackup, vm
...
+			// Remove empty string. If there is empty string in secret, we will encounter error.
+			// ref: https://github.com/harvester/harvester/issues/1536
+			data := secret.DeepCopy().Data
+			for k, v := range secret.Data {
+				if len(v) == 0 {
+					delete(data, k)
+				}
+			}
+
+			secretBackupMap[secretFullName] = true
```
Comment (guangbochen) [inline pkg/controller/master/backup/backup.go]: seems duplicated with the above check and can we move it to a method.

File: pkg/controller/master/backup/backup.go [hunk 3]
```diff
@@ -181,26 +181,56 @@ func (h *Handler) getVolumeBackups(backup *harvesterv1.VirtualMachineBackup, vm
...
 			})
 		}
 		if volume.CloudInitNoCloud != nil && volume.CloudInitNoCloud.NetworkDataSecretRef != nil {
 			secret, err := h.secretCache.Get(vm.Namespace, volume.CloudInitNoCloud.NetworkDataSecretRef.Name)
 			if err != nil {
 				return nil, err
 			}
+
+			// users may put UserDataSecretRef and NetworkDataSecretRef in a same secret, so we only keep one
+			secretFullName := fmt.Sprintf("%s/%s", secret.Namespace, secret.Name)
```
Comment (guangbochen) [inline pkg/controller/master/backup/backup.go]: should move the check to the top as RefName and vm.Namespace is same as the secret one.

File: pkg/controller/master/backup/restore.go [hunk 4]
```diff
@@ -335,29 +168,60 @@ func (h *RestoreHandler) VMOnChange(key string, vm *kv1.VirtualMachine) (*kv1.Vi
...
+		restoreCpy.Status.VolumeRestores = volumeRestores
+	}
+
+	if restoreCpy.Status.DeletedVolumes == nil {
+		var deletedVolumes []string
+		for _, vol := range backup.Status.VolumeBackups {
+			deletedVolumes = append(deletedVolumes, vol.PersistentVolumeClaim.ObjectMeta.Name)
+		}
+		restoreCpy.Status.DeletedVolumes = deletedVolumes
+	}
```
Comment (guangbochen) [inline pkg/controller/master/backup/restore.go]: probably skip adding deleted volumes of new restored VM and retained policy is preferred.

File: pkg/controller/master/backup/backup.go (hunk 2 above)
Comment (FrankYang0529) [reply pkg/controller/master/backup/backup.go]: Refactor it. Thank you.

File: pkg/controller/master/backup/restore.go (hunk 4 above)
Comment (FrankYang0529) [reply pkg/controller/master/backup/restore.go]: Check whether is the new VM or has retained policy. Thank you.

File: pkg/controller/master/backup/backup.go (hunk 3 above)
Comment (FrankYang0529) [reply pkg/controller/master/backup/backup.go]: Updated it. Thank you.

File: pkg/controller/master/backup/restore.go [hunk 5]
```diff
@@ -695,75 +551,115 @@ func (h *RestoreHandler) createRestoredPVC(vmRestore *harvesterv1.VirtualMachine
...
-	return err
+	return nil
 }
 
-func updateRestoreCondition(r *harvesterv1.VirtualMachineRestore, c harvesterv1.Condition) {
-	r.Status.Conditions = updateCondition(r.Status.Conditions, c, true)
+func (h *RestoreHandler) restartVM(vm *kv1.VirtualMachine) error {
+	logrus.Infof("restarting the vm %s, current state running:%v", vm.Name, *vm.Spec.Running)
+	var running = true
+	if vm.Spec.Running == nil || vm.Spec.Running == &running {
```
Comment (gitlawr) [inline pkg/controller/master/backup/restore.go]: `*vm.Spec.Running` instead of `vm.Spec.Running == &running`

Comment (FrankYang0529) [reply pkg/controller/master/backup/restore.go]: Fixed it. Thank you.

**Review by guangbochen — APPROVED** [review]
LGTM

**Review by gitlawr — APPROVED** [review]
LGTM

**Pull request merged**